Objective-C/CBLLiveQuery.mm

```diff
@@ -0,0 +1,179 @@
+    }
+    _willUpdate = NO; // cancels the delayed update started by -databaseChanged
+}
+
+
+- (NSArray*) rows {
+    [self start];
+    return _rows;
+}
+
+
+- (void) setRows:(NSArray*)rows {
+    _rows = rows;
+}
+
+
+- (void) addObserver:(NSObject *)observer forKeyPath:(NSString *)keyPath
+             options:(NSKeyValueObservingOptions)options context:(void *)context
+{
+    if ([keyPath isEqualToString: @"rows"]) {
+        if (++_observerCount == 1)
+            [self start];
+    }
+    [super addObserver: observer forKeyPath: keyPath options: options context: context];
+}
+
+
+- (void) removeObserver:(NSObject *)observer forKeyPath:(NSString *)keyPath {
+    if ([keyPath isEqualToString: @"rows"]) {
+        if (--_observerCount == 0)
+            [self stop];
+    }
+    [super removeObserver: observer forKeyPath: keyPath];
+}
+
+
+- (void) databaseChanged: (NSNotification*)n {
+    if (_willUpdate)
+        return;  // Already a pending update scheduled
+
+    // Use double the update interval if this is a remote change (coming from a pull replication):
+    NSTimeInterval updateInterval = _updateInterval;
+    if ([n.userInfo[kCBLDatabaseIsExternalUserInfoKey] boolValue])
+        updateInterval *= 2;
+
+    // Schedule an update, respecting the updateInterval:
+    NSTimeInterval updateDelay = (_lastUpdatedAt + updateInterval) - CFAbsoluteTimeGetCurrent();
+    updateDelay = MAX(0, MIN(_updateInterval, updateDelay));
+    [self updateAfter: updateDelay];
+}
+
+
+- (void) updateAfter: (NSTimeInterval)updateDelay {
+    if (_willUpdate)
+        return;  // Already a pending update scheduled
+    _willUpdate = YES;
+    CBLLog(Query, @"%@: Will update after %g sec...", self, updateDelay);
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(updateDelay * NSEC_PER_SEC)),
+                   dispatch_get_main_queue(), ^{        //FIX: Use a different queue
+        if (_willUpdate)
+            [self update];
+    });
+}
+
+
+- (void) update {
+    //TODO: Make this asynchronous (as in 1.x)
+    NSError *error;
+    CBLQueryEnumerator* oldEnum = _enum;
+    CBLQueryEnumerator* newEnum;
+    if (oldEnum == nil || _forceReload)
+        newEnum = (CBLQueryEnumerator*) [self run: &error];
+    else
+        newEnum = [oldEnum refresh: &error];
+
+    _willUpdate = _forceReload = false;
+    _lastUpdatedAt = CFAbsoluteTimeGetCurrent();
+
+    if (newEnum != oldEnum) {
+        CBLLog(Query, @"%@: Changed!", self);
+        _enum = newEnum;
+        self.rows = newEnum.allObjects;     // triggers KVO
+    }
+
+    if (error || _lastError) {
+        if (error)
+            CBLWarnError(Query, @"%@: Update failed: %@", self, error.localizedDescription);
+        self.lastError = error;             // triggers KVO
+    }
+}
+
+
+- (BOOL) waitForRows {
+    if (!_rows && !_lastError)
+        [self update];
+    return _rows != nil;
+}
+
+ 
+- (void) queryOptionsChanged {
```

Comment:
This is currently not in the public header and has not ben used anywhere. There is only one option which is `updateInterval`; maybe we don't need to make it public and call this method when the the updateInterval is changed?